src/MoveUntilTouchTopicController.cpp

```diff
@@ -0,0 +1,218 @@
+void MoveUntilTouchTopicController::taringTransitionCallback(TareActionClient::GoalHandle goalHandle) {
+  if (goalHandle.getResult() && goalHandle.getResult()->success) {
+    ROS_INFO("Taring completed!");
+    mTaringCompleted.store(true);
+  }
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::starting(const ros::Time& time)
+{
+  // start asynchronous tare request
+  ROS_INFO("Starting Taring");
+  pr_control_msgs::TriggerGoal goal;
+  mTareGoalHandle = mTareActionClient->sendGoal(goal, boost::bind(&MoveUntilTouchTopicController::taringTransitionCallback, this, _1));
+
+  // start base trajectory controller
+  startController(time);
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::stopping(const ros::Time& time)
+{
+  // stop base trajectory controller
+  stopController(time);
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::update(const ros::Time& time,
+                                      const ros::Duration& period)
+{
+  // update base trajectory controller
+  updateStep(time, period);
+}
+
+//=============================================================================
+bool MoveUntilTouchTopicController::shouldAcceptRequests() { return isRunning(); }
+
+//=============================================================================
+bool MoveUntilTouchTopicController::shouldStopExecution(std::string& reason)
+{
+  // inelegent to just terminate any running trajectory,
+  // but we must guarantee taring completes before starting
+  if (!mTaringCompleted.load()) {
+    ROS_WARN("taring not yet completed!");
+    return true;
+  }
+
+  double forceThreshold = mForceThreshold.load();
+  double torqueThreshold = mTorqueThreshold.load();
+
+  std::lock_guard<std::mutex> lock(mForceTorqueDataMutex);
+  bool forceThresholdExceeded = mForce.norm() >= forceThreshold;
+  bool torqueThresholdExceeded = mTorque.norm() >= torqueThreshold;
+
+  
+  if (forceThresholdExceeded) {
+    std::stringstream reasonStream;
+    reasonStream << "Force Threshold exceeded!   Threshold: " << forceThreshold << "   Force: " << mForce.x() << ", " << mForce.y() << ", " << mForce.z();
+    reason = reasonStream.str();
+    ROS_WARN(reason.c_str());
+  }
+  if (torqueThresholdExceeded) {
+    std::stringstream reasonStream;
+    reasonStream << "Torque Threshold exceeded!   Threshold: " << torqueThreshold << "   Torque: " << mTorque.x() << ", " << mTorque.y() << ", " << mTorque.z();
+    reason = reasonStream.str();
+    ROS_WARN(reason.c_str());
+  }
+
+  return forceThresholdExceeded || torqueThresholdExceeded;
+}
+
+//=============================================================================
+void MoveUntilTouchTopicController::setForceTorqueThreshold(FTThresholdGoalHandle gh)
```

Comment:
Shouldn't this be `FTThresholdGoalHandle& gh`?

Reply:
Interesting point. When I use a reference, I have trouble binding this method to the action server.